plugins/optimization-detective/class-od-url-metric-group-collection.php

```diff
@@ -396,34 +397,61 @@ public function get_common_lcp_element(): ?array {
 	}
 
 	/**
-	 * Gets the max intersection ratios of all elements across all groups and their captured URL metrics.
+	 * Gets all elements from all URL metrics from all groups keyed by the elements' XPaths.
 	 *
-	 * @return array<string, float> Keys are XPaths and values are the intersection ratios.
+	 * This is an O(n^3) function so its results must be cached. This being said, the number of groups should be 4 (one
+	 * more than the default number of breakpoints) and the number of URL metrics for each group should be 3
+	 * (the default sample size). Therefore, given the number (n) of visited elements on the page this will only
+	 * end up running n*4*3 times.
+	 *
+	 * @todo Should there be an OD_Element class which has a $url_metric property which then in turn has a $group property. Then this would only need to return array<string, OD_Element[]>.
```

Comment:
I want to address this in a follow-up PR since it will require more refactoring and I don't want to make this PR even larger.

Reply:
See https://github.com/WordPress/performance/pull/1585

Reply:
Sounds good. For the future, ideally the different plugin modifications would happen in separate PRs too. It's more convenient all in one PR, but would be better for review, plus you could still use sub-PRs like you've done before, to e.g. build something in Embed Optimizer on top of another PR change to the Optimization Detective infrastructure.

Reply:
Yes, I agree. Sorry for the large PR. It was a bit of a journey.